### ISSUE TITLE
Clarify RISC-V PLIC interrupt handling

### ISSUE THREAD

**Issue by axel-h** [seL4/seL4#464]
I'm trying to understand https://github.com/seL4/seL4/commit/375a98c8b36cea2dc3a9c8e94545a582f5c23548 where in `src/arch/riscv/machine/hardware.c` this comment was added:
```
 * On other architectures the kernel masks interrupts between delivering them to
 * userlevel and receiving the acknowledgement invocation. This strategy doesn't
 * work on RISC-V as an IRQ is implicitly masked when it is claimed, until the
 * claim is acknowledged. If we mask and unmask the interrupt at the PLIC while
 * a claim is in progress we sometimes experience IRQ sources not being masked
 * and unmasked as expected. Because of this, we don't mask and unmask IRQs that
 * are for user level, and also call plic_complete_claim for seL4_IRQHandler_Ack.
 */
```
@szhuang: On which hardware did you observe the behavior and is there an errata from the vendor about this?

What I am currently seeing is, that if interrupt claim completion does not happen within the kernel's interrupt trap handling, all interrupts can be blocked by a driver. No other interrupt arrives from the PLIC until the user mode ISR ack's the interrupt. The RISC-V PLIC specs are unclear if really no other interrupt will arrive or just interrupts or the same or lower priority are masked. I've asked in https://github.com/riscv/riscv-isa-manual/issues/689 if the spec could clarify this is bit more. Anyway, I consider this a big practical problem, we can't allow any driver to block all other interrupts. Luckily, this does not affect the timer used for the preemptive scheduler, because RISC-V has a dedicated timer interrupt that is separate from the external interrupt.

**Comment by axel-h** [comment]
I wonder if the current implementation may even allow a driver to block the scheduling timer interrupt. Assuming a claimed interrupt is not completed immediately, it remains in the `active_irq[]` list and it always presented as the current interrupt. As long as the interrupt is not completed, the timer interrupt will never be presented to the kernel when it calls `getActiveIRQ()`.

**Comment by kent-mcleod** [comment]
This comment was written according to observed behavior on the hifive unleashed.  On Arm the kernel does the following for a user level interrupt:
1. Receive interrupt in kernel
2. Signal notification object
2. Mask interrupt
3. Acknowledge IRQ with controller via EOI message
4. Return to user level
5. When user level acknowledges the interrupt, unmask the IRQ.

For the plic0 in the hifive unleashed, this same sequence caused IRQs to stop being delivered. It was necessary to remove the masking and unmasking as described in the comment.

The behavior that you are seeing seems to contradict the assumptions that were made about the PLIC, specifically that more IRQs would be delivered for different interrupt sources as multiple sources can be pending at the same time.
From the spec: `At most one interrupt request per interrupt source can be pending in the PLIC core at any time, indicated
by setting the source’s IP bit.` So you can't have 2 timer interrupts pending, but you can have one timer and one serial pending.  I could be wrong though.  It was a while ago that I last looked at this.  At the time I was using gpio interrupts and pwm timer interrupts, but I can't remember if I tested the explicit scenario you are observing (and there aren't any tests in sel4test either).

**Comment by axel-h** [comment]
Did you get in contact with SiFive about this issue to check if this is a bug or things are done incorrectly?
Has somebody tested this on the HiFive Unmatched already? I may get a board in the next weeks, until then I can only use QEMU here.

**Comment by axel-h** [comment]
It seems the RISC-V PLIC has no masking/unmasking concept. It can just enable and disable interrupts. The difference is, that a masked interrupt is not lost, but remembered and it will trigger as soon as it is unmasked. A disable interrupt is really lost. The generic kernel code seems to be build on the maskin that exists on ARM, but on RISC-V this either needs to be emulated in software then or the generic code has to change and move all this to architecture specific code. Having `plic_mask_irq()` in the API seem questionable, the implementation that accesses `enable` PLIC register is also dangerous.

**Comment by kent-mcleod** [comment]
The Arm GIC and RISC-V PLIC specs use enable/disable language for the way that the kernel masks and unmasks interrupts.  They both also describe behavior for masking with priorities, which the kernel doesn't use.
 
> The difference is, that a masked interrupt is not lost, but remembered and it will trigger as soon as it is unmasked. A disable interrupt is really lost.

What definition of masking are you referring to here? The kernel's interrupt model delivers interrupts to user level and needs to prevent them from re-triggering before the user has acknowledged.  If there is platform supported way to mask a single interrupt source then the kernel should use this. With GICs, masking can only be done via global priorities and isn't sufficient for what the kernel needs.  So it falls back to enabling and disabling - which is more expensive too.  With the GIC the kernel acknowledges the IRQ after the IRQ has been disabled and before returning to user level. The assumption is that when the interrupt is re-enabled, it will have advanced from the previous interrupt state.

With the PLIC, the spec indicates that until a claimed IRQ is acknowledged, no further interrupts of the same ID will be delivered. It doesn't specify that IRQs of different IDs wont be delivered.  If IRQs with different IDs can still be delivered then this fits the kernel's requirement for "masking" the active IRQ.  If IRQs with different IDs are not delivered then the kernel cannot rely on this as the "masking" mechanism and use a more expensive enable/disable one.  

> Assuming a claimed interrupt is not completed immediately, it remains in the active_irq[] list and it always presented as the current interrupt.

This isn't true.  The kernel set's active_irq[] to irqInvalid once the notification object has been signalled before exiting the kernel.


> Did you get in contact with SiFive about this issue to check if this is a bug or things are done incorrectly?

No.  The behavior was consistent with their manual (SiFive FU540-C000 Manual v1p0, 10.8 Interrupt Completion): `If the completion ID does not match an interrupt source that is currently enabled for the target, the completion is silently ignored.`

**Comment by axel-h** [comment]
Sorry about the `active_irq[]`, there I was indeed wrong there and mixed the clearing in `ackInterrupt()`up with `invokeIRQHandler_AckIRQ()` calles from userland. 

>  With GICs, masking can only be done via global priorities and isn't sufficient for what the kernel needs
Why is that not sufficient?

**Comment by kent-mcleod** [comment]
> Why is that not sufficient?

Because you can't use it to mask specific interrupt IDs without changing the interrupt priority of each ID.  This isn't possible to always rely on because `It is IMPLEMENTATION DEFINED whether a write to GICD_IPRIORITYR<n> changes the
priority of any active SPI.` and it's also possible for an implementation to make the priorities of some interrupt IDs fixed.
